Add support for listening to tcp handles to in the Sockets transport

davidfowl opened the pull request:
- Make ListenHandleTests shared across both transports
- Added API to signify ownership semantics of the file handle

Fixes #20842

davidfowl committed:
Add support for listening to tcp handles to in the Sockets transport
- Make ListenHandleTests shared across both transports
- Added API to signify ownership semantics of the file handle

Fixes #20842

gfoidl commented on src/Servers/Connections.Abstractions/src/FileHandleEndPoint.cs:
```diff
@@ -24,6 +25,7 @@ public FileHandleEndPoint(ulong fileHandle, FileHandleType fileHandleType)
             }
         }
 
+        public bool OwnsHandle { get; private set; }
```
```suggestion
        public bool OwnsHandle { get; }
```
I don't see any other set than in the Ctor.

gfoidl reviewed (commented):
Small nit.

Tratcher commented on src/Servers/Connections.Abstractions/src/FileHandleEndPoint.cs:
```diff
@@ -8,10 +8,11 @@ namespace Microsoft.AspNetCore.Connections
     public class FileHandleEndPoint : EndPoint
     {
-        public FileHandleEndPoint(ulong fileHandle, FileHandleType fileHandleType)
+        public FileHandleEndPoint(ulong fileHandle, FileHandleType fileHandleType, bool ownsHandle = true)
```
The binary breaking change guidelines say this should be a new overload rather than an optional parameter.

davidfowl replied:
Yea I wasn't sure about that. Will change it.

Tratcher commented on src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs:
```diff
@@ -147,6 +153,8 @@ public ValueTask UnbindAsync(CancellationToken cancellationToken = default)
 
         public ValueTask DisposeAsync()
         {
+            _socketHandle?.Dispose();
```
Would it be better to dispose the Socket first so it's not trying to utilize an already disposed handle?

davidfowl replied:
Done

davidfowl committed:
Few changes
- PR feedback
- Don't Bind to the endpoint

davidfowl committed:
Remove test

davidfowl committed:
Update src/Servers/Connections.Abstractions/src/FileHandleEndPoint.cs

Co-authored-by: Günther Foidl <gue@korporal.at>

davidfowl commented:
@Tratcher @halter73 this is good to go.

cc @dotnet/aspnet-api-review can I get a review

halter73 commented on src/Servers/Connections.Abstractions/ref/Microsoft.AspNetCore.Connections.Abstractions.netcoreapp.cs:
```diff
@@ -100,8 +100,10 @@ public partial class DefaultConnectionContext : Microsoft.AspNetCore.Connections
+        public FileHandleEndPoint(ulong fileHandle, Microsoft.AspNetCore.Connections.FileHandleType fileHandleType, bool ownsHandle) { }
         public ulong FileHandle { [System.Runtime.CompilerServices.CompilerGeneratedAttribute] get { throw null; } }
         public Microsoft.AspNetCore.Connections.FileHandleType FileHandleType { [System.Runtime.CompilerServices.CompilerGeneratedAttribute] get { throw null; } }
+        public bool OwnsHandle { [System.Runtime.CompilerServices.CompilerGeneratedAttribute] get { throw null; } }
```
The libuv transport always closes the FileHandleEndPoint when it unbinds. If we're going to add this API, we should probably have libuv respect it.

A test to verify there's really a difference in behavior would also be nice.

davidfowl replied:
You want me to do extra work? OK

halter73 replied:
If it's really too difficult, you could throw a NotSupportedException or something from libuv when OwnsHandle is false.

davidfowl replied:
I think that’s what I’ll do

davidfowl committed:
Throw an exception if libuv does not own the handle

davidfowl commented:
@halter73 update?

halter73 commented:
Can you move ListenHandleTests to Kestrel's main shared FunctionalTests directory so it's run as part of Sockets.FunctionalTests in addition to Libuv.FunctionalTests?

We should also add a version of [CanListenToOpenTcpSocketHandle](https://github.com/dotnet/aspnetcore/blob/40e0649785feafc6c59337492eb1c5c5c081d663/src/Servers/Kestrel/test/Libuv.FunctionalTests/ListenHandleTests.cs#L25) which sets OwnsHandle to false and verifies the server can successfully be started and stopped with the same handle twice. Otherwise, we have no test coverage for the new OwnsHandle API.

davidfowl commented:
> Can you move ListenHandleTests to Kestrel's main shared FunctionalTests directory so it's run as part of Sockets.FunctionalTests in addition to Libuv.FunctionalTests?

It's already moved and running for both.

> We should also add a version of CanListenToOpenTcpSocketHandle which sets OwnsHandle to false and verifies the server can successfully be started and stopped with the same handle twice. Otherwise, we have no test coverage for the new OwnsHandle API.

I'll add this test to the socket transport tests.

davidfowl committed:
Add test for owns handle
- Make sure we can bind multiple times with the same handle if owns handle is false.
- Make sure it throws on libuv

davidfowl committed:
Removed ownshandle
- This didn't work properly and prevented the accept loop from properly yielding when disposing the socket.

davidfowl commented:
OK @halter73 I had to remove OwnsHandle because there's some weird behavior without it. The issue is that the accept loop doesn't properly yield when you dispose the listen socket. Not worth chasing down in this PR, I'll start a discussion with @stephentoub on dotnet runtime with an isolated repro.
